### ISSUE TITLE
Kotlin Cleanup

### ISSUE THREAD

**Issue by Arthur-Milchior** [ankidroid/Anki-Android#10489]
Currently, there are 84 `@KotlinCleanup` in AnkiDroid code source. Most of those changes are supposed to be simple to do and would be a good pull request by themselves. Also, if you are a java developer and new to Kotlin, it will be a great way to learn Kotlin advantages, since it's usually tasks that were introduced when translating java to Kotlin, and which can actually be improved in Kotlin compared to original java code.

Ideally, each PR should contains a single cleanup. This ensure the PR is very very quick to review and merge. The main exception would be if a clean-up is dependent on another clean-up and this would cause merge conflict.

**Comment by kilitr** [comment]
i would like to work on this

**Comment by david-allison** [comment]
Sounds good! IMO it's not worth assigning to this one as there's so much to do.

Ideally one PR per class - just let us know which one you're working on so someone else doesn't work on it at the same time

**Comment by kilitr** [comment]
### Working on CustomStudyDialog.kt

I would like some help concerning the second part of this KotlinCleanup
https://github.com/ankidroid/Anki-Android/blob/a9ca2cfa854d999e5e4cbe905239b9109f1aee9c/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt#L343-L417

What exactly is meant by "mapping dialogID to an enum"?

1. IMO, the only thing that would make sense is replacing the CONTEXT_MENU_* and all other fields of the companion object (see below) with two seperate enums (One for the context menu configuration and one for the context menu options). However this would be a huge change across this file.
2. Introducing an additional enum which dialogID will get mapped to before entering the when statement would be the alternative. This feels dirty to me due to code duplication

https://github.com/ankidroid/Anki-Android/blob/a9ca2cfa854d999e5e4cbe905239b9109f1aee9c/AnkiDroid/src/main/java/com/ichi2/anki/dialogs/customstudy/CustomStudyDialog.kt#L500-L520

**Comment by divyansh-dxn** [comment]
I would also like to work on it. So basically, I have to go through each class marked with @KotlinCleanup and convert ko Kotlin making sure that changes don't break and pass all the tests.

**Comment by Arthur-Milchior** [comment]
> I would also like to work on it. So basically, I have to go through each class marked with @KotlinCleanup and convert ko Kotlin making sure that changes don't break and pass all the tests.

Well, you just need to find one `@KotlinCleanup` that is not the one @kilitr is working on, and correct it. Please don't try to solve all clean up in a single PR, that would be very hard to review

**Comment by Arthur-Milchior** [comment]
@kilitr Please do the thing that makes sens to you. 
This guidance is true all of the time actually. It's easy to be not clear, to miss a nuance, while writtings todo/issues, that is only really catched while implementing the code. It's quite rare that we require something to be done exactly the way we required it, and it never occurs that we are not willing to discuss our demands. So, if you do the work and that you find an idea that is great, feel free to do it. We can always discuss it in details during the code review. 
The only case where I would really want you to ask for permission before doing the work is if the work will take you hours; and it's only because that will avoid you to write code during hours and then lose all of this works when the PR gets closed without being merged

**Comment by divyansh-dxn** [comment]
Currently I'm working on `DeckPickerFloatingActionMenu.kt`
Other working on this issue, please mention which class/file they are working on.

**Comment by DeepanshuPratik** [comment]
Currently, I'm working on `CardTemplatePreviewer.kt`.
Others also please mention which files they are working on to avoid merge conflicts.

**Comment by david-allison** [comment]
> Currently, I'm working on `CardTemplatePreviewer.kt`. Others also please mention which files they are working on to avoid merge conflicts.

Just FYI: conflicts with https://github.com/ankidroid/Anki-Android/pull/10526

**Comment by DeepanshuPratik** [comment]
> > Currently, I'm working on `CardTemplatePreviewer.kt`. Others also please mention which files they are working on to avoid merge conflicts.
> 
> Just FYI: conflicts with #10526

But that change is done for only one variable `mCardList` but two more variables need to be updated, can you please verify?

**Comment by DeepanshuPratik** [comment]
> > Currently, I'm working on `CardTemplatePreviewer.kt`. Others also please mention which files they are working on to avoid merge conflicts.
> 
> Just FYI: conflicts with #10526

Thanks, I didn't notice it.

**Comment by DeepanshuPratik** [comment]
Currently, I'm working on `LanguageUtils.kt`.

**Comment by TheScriptan** [comment]
Heya, I'll take on cleaning up `ImportUtils.kt`

**Comment by Bhawna1203** [comment]
@Arthur-Milchior @mikehardy. Sir can i work on AsyncDialogFragment,kt

**Comment by david-allison** [comment]
> @Arthur-Milchior @mikehardy. @david-allison  Sir can i work on AsyncDialogFragment,kt

Sure, and no need for the "sir" 😇

**Comment by Jatin887** [comment]
I also want to work on it . can you please assign it to me

**Comment by david-allison** [comment]
> Sounds good! IMO it's not worth assigning to this one as there's so much to do and it'd be great to get other people on this at the same time
>
> Ideally one PR per class - just let us know which one you're working on so someone else doesn't work on it at the same time

https://github.com/ankidroid/Anki-Android/issues/10489#issuecomment-1065610715

**Comment by Yugal41735** [comment]
Can anyone tell me what exactly I have to do in this issue?

**Comment by kilitr** [comment]
> Can anyone tell me what exactly I have to do in this issue?

Search the entire codebase for `@KotlinCleanup("Description")` annotations... read and understand the description and resolve whatever issue was pointed out in the Description.

One basic example that you'll come across very often is remove nullability from a certain Field.

**Comment by Yugal41735** [comment]
![image](https://user-images.githubusercontent.com/74638775/161550232-eeb26bc7-82ae-49e9-8cd0-922dfb317f5b.png)

So for example: in the above codebase what exactly I have to do?

**Comment by Yugal41735** [comment]
anyone could explain what exactly I have to do in the above codebase?

**Comment by Umesh-01** [comment]
@david-allison Is there anything left in this issue?? I would like to work on it.

**Comment by raymondwzeng** [comment]
Started tackling `MyAccount.kt`'s various cleanup tasks. Might have a PR in later once I verify that the tests pass.

**Comment by Jatin887** [comment]
I am facing this issue while setting up the project can you plese help me out and i wanna convert Taskmangaer.java async task to kotlin coroutines as i am intrested in the project async task to coroutines 
![image](https://user-images.githubusercontent.com/71597104/162071534-38607f96-fb55-4811-b2ed-865531a48f5a.png)

**Comment by david-allison** [comment]
> I am facing this issue while setting up the project can you plese help me out and i wanna convert Taskmangaer.java async task to kotlin coroutines as i am intrested in the project async task to coroutines 
> ![image](https://user-images.githubusercontent.com/71597104/162071534-38607f96-fb55-4811-b2ed-865531a48f5a.png)
> 

What version of Android Studio are you using?

**Comment by david-allison** [comment]
> @david-allison Is there anything left in this issue?? I would like to work on it.

Lots to do - just state the class you're working on so you don't get conflicts

**Comment by arpitmx** [comment]
Heya👋!
I am currently working on Kotlin Cleanup for `Counts.kt`.  

File link :  [https://github.com/ankidroid/Anki-Android/blob/main/AnkiDroid/src/main/java/com/ichi2/libanki/sched/Counts.kt](url)

**Comment by Jatin887** [comment]
> > I am facing this issue while setting up the project can you plese help me out and i wanna convert Taskmangaer.java async task to kotlin coroutines as i am intrested in the project async task to coroutines
> > ![image](https://user-images.githubusercontent.com/71597104/162071534-38607f96-fb55-4811-b2ed-865531a48f5a.png)
> 
> What version of Android Studio are you using?

![image](https://user-images.githubusercontent.com/71597104/162139043-4cfcd63b-6254-4a88-a9d2-ff09dc2f8ae2.png)

**Comment by Jatin887** [comment]
it is showing this error but i already set the sdk path in local.properties
![image](https://user-images.githubusercontent.com/71597104/162150049-7c567bfc-3af3-42a5-963e-f42f7db3984e.png)

**Comment by Umesh-01** [comment]
Currently, I am working on Kotlin Cleanup for `HelpDialog.kt`

**Comment by Umesh-01** [comment]
@david-allison it is showing one error in `Preferences.kt` file when I opened this project, do I have to import something??

![image](https://user-images.githubusercontent.com/83420185/162256568-9f8b39d7-9529-40ad-89e0-dff1413111fc.png)

**Comment by david-allison** [comment]
> @david-allison it is showing one error in `Preferences.kt` file when I opened this project, do I have to import something??
> 
> ![image](https://user-images.githubusercontent.com/83420185/162256568-9f8b39d7-9529-40ad-89e0-dff1413111fc.png)
> 

You need to build the project

**Comment by Jatin887** [comment]
@david-allison i converted CompatV21.java to CompatV21.kt 
The project is working as it should be 
can i go ahead and create a pr for the same?

**Comment by Umesh-01** [comment]
> > @david-allison it is showing one error in `Preferences.kt` file when I opened this project, do I have to import something??
> 
> You need to build the project

Yes I did, and it is working fine when I used this import `import com.arcao.slf4j.timber.BuildConfig`

**Comment by UvrajSB** [comment]
I am working on Preferences.kt  { @KotlinCleanup("iteration with i") }

**Comment by david-allison** [comment]
> Yes I did, and it is working fine when I used this import `import com.arcao.slf4j.timber.BuildConfig`

That import is incorrect. You want the one from the `build` folder which is used once the project has been built.

**Comment by UvrajSB** [comment]
I am taking up the SeekBarPrefence.kt

**Comment by SySagar** [comment]
> I am taking up the SeekBarPrefence.kt

Can I take it please?

**Comment by UvrajSB** [comment]
> > I am taking up the SeekBarPrefence.kt
> 
> Can I take it please?

Sorry, I have started working on this, can you kindly look for some other part.

**Comment by UvrajSB** [comment]
@david-allison what do "autoLint" mean, I understand meaning of Lint but in this file there is cleanup by the name autoLint. What am I really asked to do by that?

**Comment by UvrajSB** [comment]
Also, Do you want me to create one PR for one file or can I make one PR for multiple files as somefiles even have a single kotlinCleanup?

**Comment by david-allison** [comment]
I understand meaning of Lint but in this file there is cleanup by the name autoLint. What am I really asked to do by that?

There's a lot of automatic lint fixes that Android Studio suggests

**Comment by david-allison** [comment]
> Also, Do you want me to create one PR for one file or can I make one PR for multiple files as somefiles even have a single kotlinCleanup?

Ideally one PR per file, definitely one commit per file

**Comment by UvrajSB** [comment]
@david-allison 
![image](https://user-images.githubusercontent.com/73571511/162623517-6945c835-5676-4d1e-8a2f-ce26e9768976.png)
why these variables are initialized in init block? android studio is giving the suggestion to initialize them with the declaration.

**Comment by david-allison** [comment]
That's what the automatic Kotlin migration did

**Comment by Jatin887** [comment]
![image](https://user-images.githubusercontent.com/71597104/162640390-4d608d34-23dc-4467-992b-d039e0433eb3.png)
What does this actually mean . i just have to convert all the Supress into SupressLint right?

**Comment by UvrajSB** [comment]
> ![image](https://user-images.githubusercontent.com/71597104/162640390-4d608d34-23dc-4467-992b-d039e0433eb3.png) What does this actually mean . i just have to convert all the Suppress into SupressLint right?

Just right-click anywhere inside the file and then select analyze > Inspect code 
and then Android studio will give suggestions, try to implement those.
(By unused classes perhaps mean that you need not to do any changes in classes which are not used)

**Comment by UvrajSB** [comment]
I am trying to resolve PreferencesTest.kt migration

**Comment by UvrajSB** [comment]
I am working on KeyboardShortcutIntegrationTest

**Comment by UvrajSB** [comment]
I am unsure about the use of scope function here, can anyone please explain
![image](https://user-images.githubusercontent.com/73571511/163651151-97803ae4-3af6-4828-a6d1-c918225ae198.png)

**Comment by Jatin887** [comment]
I am working on RecylerSingleTouchAdapter.kt

**Comment by Jatin887** [comment]
I am worling on LoadPronunciationActivity.kt

**Comment by david-allison** [comment]
I'm working on some of `DeckPicker.kt`

**Comment by david-allison** [comment]
`AnalyticsTest.kt`

**Comment by UvrajSB** [comment]
I will work on JSONObjectTest once this PR is merged
https://github.com/ankidroid/Anki-Android/pull/10933#issue-1207130014

**Comment by Sustech-yx** [comment]
I'm working on CardBrowser.kt

**Comment by UvrajSB** [comment]
I am taking CheckBoxTriStates.kt

**Comment by UvrajSB** [comment]
@david-allison Do we really need to use ` ?.let ` here ??

<img src = "https://user-images.githubusercontent.com/73571511/165459497-0ba7c390-04f2-43eb-87fe-fae9091e2e6e.jpeg" width = "350" height = "250"/>

because it can be easily done like this and the let would have been redundant in this case




![WhatsApp Image 2022-04-27 at 12 25 35 PM](https://user-images.githubusercontent.com/73571511/165459664-8fc5c618-1e08-43d8-acb6-c83adb150d6a.jpeg)

**Comment by Jatin887** [comment]
I am working on TranslationActivity.kt

**Comment by Jatin887** [comment]
I am working on OverviewStatusBuilder.kt

**Comment by david-allison** [comment]
> @david-allison Do we really need to use `?.let` here ??
> 
> ![WhatsApp Image 2022-04-27 at 12 15 06 PM](https://user-images.githubusercontent.com/73571511/165460358-cfcc0e78-a223-4272-992a-3396ae09200a.jpeg)
> 
> because it can be easily done like this and the let would have been redundant in this case
> 
> ![WhatsApp Image 2022-04-27 at 12 25 35 PM](https://user-images.githubusercontent.com/73571511/165459664-8fc5c618-1e08-43d8-acb6-c83adb150d6a.jpeg)

`?.` is fine here

**Comment by vinaydeepkaur** [comment]
Currently working on VideoPlayer.kt

**Comment by suman-somu** [comment]
I am working on IField.kt

**Comment by Jatin887** [comment]
I am working on AsyncDialogFragment.kt

**Comment by Jatin887** [comment]
I am working on ActivityExportingDelegate.kt

**Comment by Arthur-Milchior** [comment]
Thanks a lot for all of the cleanup already done. 26 of them is already impressive.

I updated the first message to add a requirement:

> Ideally, each PR should contains a single cleanup. This ensure the PR is very very quick to review and merge. The main exception would be if a clean-up is dependent on another clean-up and this would cause merge conflict.

**Comment by bennettj12** [comment]
I'm currently working on NavigationDrawerActivity.kt

**Comment by divyansh-dxn** [comment]
Keep open

**Comment by sudhanshu1309** [comment]
I am working on AbstractFlashcardViewer

**Comment by alyoshenka** [comment]
I am working on `FieldEditText.kt`

**Comment by TimSta82** [comment]
I am working on `ContentProviderTest`

**Comment by Ash-2k3** [comment]
I am working on `TemporaryModel.kt`.
In this particular class, there are couple of `@KotlinCleanup` mentions.
Do we have to include all in one PR or just one at a time ?

**Comment by rishujam** [comment]
Hey I would like to work `AbstractFlashcardViewer.kt` file. Can I?

**Comment by rishujam** [comment]
<img width="834" alt="Screenshot 2023-02-09 at 1 16 10 AM" src="https://user-images.githubusercontent.com/74773876/217635748-437adad4-83cc-46c3-9743-58b6b224e1ad.png">
can't understand what's wrong in this everything looks fine. The variable is used in the inner class SelectEaseHandler.

**Comment by VENKATAKUMARA** [comment]
I would like to work on this issue and as I am new to open source I would like to know more about this issue !

**Comment by david-allison** [comment]
Closing this as WONTFIX [as proposed on Discord](https://discord.com/channels/368267295601983490/701922522836369498/1075360932299030639) in favour of:

- https://github.com/ankidroid/Anki-Android/issues/13283
- https://github.com/ankidroid/Anki-Android/issues/13282

----

The `@KotlinCleanup` annotation is useful for guidance when refactoring the code, and I would like to see the annotation removed from the code eventually. 

But, as a 'good first issue', it produced a large number of contributions which were both high risk, and low reward.

These took up a large proportion of reviewer time and were often in critical areas of the codebase. By nature of being a 'good first issue', these changes are open to users who are unfamiliar with our codebase and may be unfamiliar with Android. We shouldn't be guiding these contributors to high risk areas of the code, especially for low reward changes.

I'd prefer to see efforts go into our open issue list going forwards.
